File: apps/android/app/src/test/kotlin/com/loonext/android/core/dashboard/DashboardPanelsTest.kt
package com.loonext.android.core.dashboard

import java.io.File
import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertTrue
import org.junit.Test

/**
 * #540 — which parts of the landing screen a member may put away.
 *
 * Two halves. The behaviour tests assert the rules; the parity tests read the
 * shared TypeScript, because this is a hand-port and nothing about Kotlin says
 * the original moved.
 */
class DashboardPanelsTest {

    @Test
    fun `nothing is hidden by default`() {
        // The other direction — an opt-in dashboard — means a new card is
        // invisible to every existing member forever, which is how a feature
        // ships to nobody.
        assertEquals(emptyList<DashboardPanels.Panel>(), DashboardPanels.normalise(emptyList()))
        for (panel in DashboardPanels.Panel.entries) {
            assertTrue(DashboardPanels.isVisible(emptyList(), panel))
        }
    }

    @Test
    fun `an id this build does not know is dropped, not an error`() {
        // A server ahead of this app, or a card we withdrew. The member gets a
        // working dashboard showing one panel they had put away — recoverable in
        // a tap — rather than a crash where their screen used to be.
        assertEquals(
            listOf(DashboardPanels.Panel.PIPELINE),
            DashboardPanels.normalise(listOf("pipeline", "crystal_ball")),
        )
    }

    @Test
    fun `duplicates collapse`() {
        assertEquals(
            listOf(DashboardPanels.Panel.PIPELINE),
            DashboardPanels.normalise(listOf("pipeline", "pipeline")),
        )
    }

    @Test
    fun `the order is the declared one, not the tapping order`() {
        assertEquals(
            listOf(DashboardPanels.Panel.RESPONSE_TIME, DashboardPanels.Panel.RECENT_CALLS),
            DashboardPanels.normalise(listOf("recent_calls", "response_time")),
        )
    }

    @Test
    fun `a hidden panel is not visible and its neighbours still are`() {
        val hidden = listOf("pipeline")
        assertFalse(DashboardPanels.isVisible(hidden, DashboardPanels.Panel.PIPELINE))
        assertTrue(DashboardPanels.isVisible(hidden, DashboardPanels.Panel.SATISFACTION))
    }

    @Test
    fun `every panel has a name and a reason`() {
        // A switch with no label is a switch nobody touches, and a name with no
        // reason is a guess for anybody who has not read both cards it could mean.
        for (panel in DashboardPanels.Panel.entries) {
            assertTrue(DashboardPanels.label(panel).length > 2)
            assertTrue(DashboardPanels.note(panel).endsWith("."))
        }
    }

    @Test
    fun `no queue section is offered as hideable`() {
        // THE LINE. Hiding unclaimed work is not a preference — it is a way to
        // stop seeing leads nobody has answered.
        val panelIds = DashboardPanels.Panel.entries.map { it.id }
        for (tile in DashboardTiles.Tile.entries) {
            assertFalse(
                "${tile.name} must never be hideable",
                panelIds.contains(tile.name.lowercase()),
            )
        }
    }

    // ---------------------------------------------------- against the original

    private fun repoFile(relative: String): String {
        var dir: File? = File("").absoluteFile
        while (dir != null) {
            val candidate = File(dir, relative)
            if (candidate.exists()) return candidate.readText()
            dir = dir.parentFile
        }
        throw AssertionError("$relative not found from ${File("").absolutePath}")
    }

    /**
     * The same five ids, in the same order, as the shared module.
     *
     * These are STORED and sent to the server, so a drift here is not a cosmetic
     * difference — it is a phone writing a preference the Worker rejects, or
     * silently dropping one the laptop saved.
     */
    @Test
    fun `the panel ids match the shared module, in order`() {
        val shared = repoFile("packages/shared/src/dashboard-panels.ts")
        val declared = Regex("""export const DASHBOARD_PANEL_IDS = \[([^\]]+)\]""")
            .find(shared)
            ?.groupValues
            ?.get(1)
            ?: throw AssertionError("DASHBOARD_PANEL_IDS is no longer an array literal")
        val ids = Regex(""""([a-z_]+)"""").findAll(declared).map { it.groupValues[1] }.toList()
        assertEquals(ids, DashboardPanels.Panel.entries.map { it.id })
    }

    /**
     * And the labels match, because they are what a member reads.
     *
     * A crew comparing a laptop and a phone over a van bonnet is comparing these
     * exact words; "Lead sources" here against "Where customers came from" there
     * reads as two different settings.
     */
    @Test
    fun `the panel labels match the shared module`() {
        val shared = repoFile("packages/shared/src/dashboard-panels.ts")
        for (panel in DashboardPanels.Panel.entries) {
            val label = DashboardPanels.label(panel)
            assertTrue(
                "the label for ${panel.id} has drifted from the shared module: $label",
                shared.contains("${panel.id}: \"$label\""),
            )
        }
    }
}

File: packages/shared/src/dashboard-panels.test.ts
import { describe, expect, it } from "vitest";

import {
  DASHBOARD_PANEL_IDS,
  DASHBOARD_PANEL_LABELS,
  DASHBOARD_PANEL_NOTES,
  DASHBOARD_PANELS_DEFAULT,
  isPanelVisible,
  normaliseHiddenPanels,
} from "./dashboard-panels";
import { DASHBOARD_TILE_LABELS } from "./dashboard-tiles";
import { EN as WEB_EN, FR_CA as WEB_FR } from "../../../apps/web/src/i18n/catalog";

/*
 * #228 — the tables above name catalogue keys, so the copy assertions resolve
 * them through the catalogue the switches actually read.
 */
function look(table: unknown, key: string): string {
  const [section, name] = key.split(".");
  const value = (table as Record<string, Record<string, string>>)[section]?.[
    name
  ];
  if (typeof value !== "string") throw new Error(`no entry for ${key}`);
  return value;
}

describe("dashboard panels (#540)", () => {
  it("hides nothing by default", () => {
    // A new member gets the whole screen and takes things off it. The other
    // direction — an opt-in dashboard — means a new card is invisible to every
    // existing member forever, which is how a feature ships to nobody.
    expect(DASHBOARD_PANELS_DEFAULT).toEqual([]);
  });

  it("offers no queue section as hideable", () => {
    // THE LINE. Hiding unclaimed work is not a preference — it is a way to stop
    // seeing leads nobody has answered. If a tile id ever appears in the panel
    // list this test is the thing that says so.
    for (const tile of Object.keys(DASHBOARD_TILE_LABELS)) {
      expect(DASHBOARD_PANEL_IDS as readonly string[]).not.toContain(tile);
    }
  });

  it("drops an id it no longer renders instead of refusing the set", () => {
    // A client one release behind, or a card we withdrew. The member gets a
    // working dashboard showing one panel they had put away — recoverable in a
    // tap — rather than an error where their screen used to be.
    expect(normaliseHiddenPanels(["pipeline", "crystal_ball"])).toEqual([
      "pipeline",
    ]);
  });

  it("collapses duplicates", () => {
    expect(normaliseHiddenPanels(["pipeline", "pipeline"])).toEqual(["pipeline"]);
  });

  it("stores the declared order, not the clicking order", () => {
    // Otherwise the same screen has several stored spellings and nothing that
    // compares two sets can be trusted.
    expect(normaliseHiddenPanels(["recent_calls", "response_time"])).toEqual([
      "response_time",
      "recent_calls",
    ]);
  });

  it("survives an empty and a fully hidden set", () => {
    expect(normaliseHiddenPanels([])).toEqual([]);
    expect(normaliseHiddenPanels([...DASHBOARD_PANEL_IDS])).toEqual([
      ...DASHBOARD_PANEL_IDS,
    ]);
  });

  it("reports a panel visible unless it is in the set", () => {
    expect(isPanelVisible([], "pipeline")).toBe(true);
    expect(isPanelVisible(["pipeline"], "pipeline")).toBe(false);
    expect(isPanelVisible(["pipeline"], "satisfaction")).toBe(true);
  });

  it("names and explains every panel", () => {
    // A switch with no label is a switch nobody touches, and a missing entry
    // would render as an empty row rather than as a crash. Since #228 these
    // tables hold catalogue KEYS, so the check is that every id has one on
    // both sides — the words themselves are checked below, in both languages.
    for (const id of DASHBOARD_PANEL_IDS) {
      expect(DASHBOARD_PANEL_LABELS[id]?.length).toBeGreaterThan(2);
      expect(DASHBOARD_PANEL_NOTES[id]?.length).toBeGreaterThan(2);
    }
    expect(Object.keys(DASHBOARD_PANEL_LABELS).sort()).toEqual(
      [...DASHBOARD_PANEL_IDS].sort(),
    );
    expect(Object.keys(DASHBOARD_PANEL_NOTES).sort()).toEqual(
      [...DASHBOARD_PANEL_IDS].sort(),
    );
  });

  it("reads in both languages (#228)", () => {
    for (const id of DASHBOARD_PANEL_IDS) {
      const label = DASHBOARD_PANEL_LABELS[id];
      const note = DASHBOARD_PANEL_NOTES[id];
      for (const [language, table] of [
        ["English", WEB_EN],
        ["French", WEB_FR],
      ] as const) {
        expect(look(table, label), `${language} for ${label}`).not.toBe("");
        // A sentence, so it ends like one — the notes sit under the names and a
        // fragment there reads as truncation. True of the French too.
        expect(look(table, note), `${language} for ${note}`).toMatch(/\.$/);
      }
      // Only the notes. A one-word label can legitimately be identical in
      // both languages — "Satisfaction" is the French for it — so asserting a
      // difference there would be asserting a worse translation. A whole
      // sentence coinciding is always the English left in place.
      expect(look(WEB_FR, note), `${note} is not translated`).not.toBe(
        look(WEB_EN, note),
      );
    }
  });

  it("gives each panel its own name", () => {
    // Two switches reading the same word is a picker you cannot use, and it
    // would pass every assertion above.
    const names = DASHBOARD_PANEL_IDS.map((id) =>
      look(WEB_EN, DASHBOARD_PANEL_LABELS[id]),
    );
    expect(new Set(names).size).toBe(DASHBOARD_PANEL_IDS.length);
  });
});
